Non-copyable generics fixes

slavapestov opened the pull request:
This fixes some validation test failures when non-copyable generics are enabled. Should be NFC otherwise.

slavapestov committed:
Sema: Introduce ResolveValueWitnessesRequest

slavapestov committed:
Sema: Refactor ResolveImplicitMemberRequest to not call TypeChecker::checkConformance()

slavapestov committed:
Sema: Refactor evaluateMembersRequest() to not call TypeChecker::checkConformance()

slavapestov committed:
Sema: Refactor computeLoweredProperties() to not call TypeChecker::checkConformance()

slavapestov committed:
Sema: Remove TypeChecker::checkConformance()

slavapestov committed:
Sema: We might be generic but not have parsed generic parameters

slavapestov committed:
IDE: Skip inverse protocols in CodeCompletionResultType's walk

slavapestov committed:
AST: Refactor ASTPrinter to use getRequirementsWithInverses()

slavapestov committed:
SIL: Add some assertions

slavapestov commented:
@swift-ci Please smoke test

slavapestov committed:
Sema: Debug messages for TypeCheckCircularity.cpp

slavapestov committed:
Sema: We can end up in deriveConformanceForInvertible() with a NominalTypeDecl from a SynthesizedFileUnit

slavapestov committed:
Distributed: Fixes for non-copyable generics

slavapestov committed:
ClangImporter: Fixes for non-copyable generics

slavapestov committed:
AST: Introduce GenericSignature::getRequirementsWithInverses()

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov committed:
AST: Refactor the mangler to use getRequirementsWithInverses()

slavapestov committed:
Sema: Move -debug-generic-signatures behavior to TypeCheckDeclPrimary.cpp

slavapestov committed:
ASTPrinter: Refactor printing of RequirementSignatures

slavapestov committed:
SILGen: Clean up substitution map construction to deal with Copyable

slavapestov committed:
Sema: Simplify DerivedConformance::createBuiltinCall()

kavon commented on lib/Sema/TypeCheckDeclPrimary.cpp:
```diff
@@ -2930,6 +2955,10 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
   void visitEnumDecl(EnumDecl *ED) {
     checkUnsupportedNestedType(ED);
 
+    // Force creation of the generic signature.
+    (void) ED->getGenericSignature();
+    dumpGenericSignature(Ctx, ED);
+
```
I'm curious why we have to do this now.

slavapestov replied:
We force certain requests in TypeCheckDeclPrimary just to ensure diagnostics are emitted before we go on to SILGen.

slavapestov committed:
Sema: Implement new behavior of extensions with non-copyable generics

We want extensions to introduce default Copyable/Escapable just like
other generic contexts, so that once Optional adopts ~Copyable,
an `extension Optional` actually adds `Wrapped: Copyable` by default.

ahoppen commented on lib/IDE/CodeCompletionResultType.cpp:
```diff
@@ -234,21 +245,33 @@ const USRBasedType *USRBasedType::fromType(Type Ty, USRBasedTypeArena &Arena) {
+          // USR mangling produces the type 'Any' for the protocol type.
           continue;
         }
+        LLVM_DEBUG(llvm::dbgs() << "Adding conformed protocol "
+                                << Conformance->getProtocol()->getName()
+                                << "\n";);
```
Are these debug statements intentionally left?

slavapestov replied:
They're only active when you run with `-Xllvm -debug-only=CodeCompletionResultType`

ahoppen commented on lib/IDE/CodeCompletionResultType.cpp:
```diff
@@ -22,6 +22,9 @@ using namespace swift;
 using namespace ide;
 using TypeRelation = CodeCompletionResultTypeRelation;
 
+#define DEBUG_TYPE "CodeCompletionResultType"
```
Stupid question: What is `DEBUG_TYPE` used for?

ahoppen commented on lib/IDE/CodeCompletionResultType.cpp:
```diff
@@ -217,13 +220,21 @@ const USRBasedType *USRBasedType::fromType(Type Ty, USRBasedTypeArena &Arena) {
         if (Proto != inherited &&
-            !inherited->isSpecificProtocol(KnownProtocolKind::Sendable)) {
+            !inherited->isSpecificProtocol(KnownProtocolKind::Sendable) &&
+            !inherited->getInvertibleProtocolKind()) {
```
I’m not sure how the inverted protocols are modeled in the AST but shouldn’t we check if the specific conformance is inverted (ie. has `~`) instead of checking whether the protocol is invertible?

slavapestov replied:
We're just skipping these protocols entirely here because we get the wrong USRs for their ProtocolDecls. Once the USR mangling is fixed we can remove this hack.

ahoppen commented on lib/IDE/CodeCompletionResultType.cpp:
```diff
@@ -234,21 +245,33 @@ const USRBasedType *USRBasedType::fromType(Type Ty, USRBasedTypeArena &Arena) {
           // get different results for its conformance. For now, always drop the
           // Sendable conformance.
+          //
+          // FIXME: Copyable and Escapable conformances are skipped because the
+          // USR mangling produces the type 'Any' for the protocol type.
```
Why does it produce `Any` for the protocol name?

slavapestov replied:
Because we need to fix the mangling.

ahoppen replied:
Ah, I see. Thanks for the clarification. Is the name mangling something we are planning to fix in the short term?

ahoppen replied:
OK, great 👍🏽

ahoppen reviewed (commented):
Only reviewed the code completion parts.

slavapestov committed:
AST: Remove ProtocolDecl::getRequirementSignatureAsGenericSignature()

slavapestov commented:
@swift-ci Please smoke test
